Kernel/FileSystem/Ext2Base.h

```diff
@@ -0,0 +1,312 @@
+    u16 raid_stride;               // RAID stride
+    u16 mmp_interval;              // # seconds to wait in MMP checking
+    u64 mmp_block;                 // Block for multi-mount protection
+    u32 raid_stripe_width;         // blocks on all data disks (N*stride)
+    u8  log_groups_per_flex;       // FLEX_BG group size
+    u8  reserved_char_pad;
+    u16 reserved_pad;  // Padding to next 32bits
+    u32 reserved[162]; // Padding to the end of the block
+    
+    InodeIndex first_inode() const {
+        if ((Revision)rev_level == Revision::Original)
+            // Have to do this ugly cast
+            return (u64)Ext2::InodeNums::FirstInode;
+        else
+            return first_inode_num;
+    }
+    size_t inode_size() const {
+        if ((Revision)rev_level == Revision::Original)
+            return good_old_inode_size;
+        else
+            return inode_size_num;
+    }
+    size_t block_size() const {
+        return (1 << min_block_log_size) << block_log_size;
+    }
+    size_t fragment_size() const {
+        return (1 << min_fragment_log_size) << fragment_log_size;
+    }
+    // Returns the exponent of the block's size.
+    size_t block_size_bits() const {
+        return block_log_size + min_block_log_size;
+    }
+    size_t inodes_per_block() const {
+        return block_size() / inode_size();
+    }
+    size_t fragments_per_block() const {
+        return block_size() / fragment_size();
+    }
+    size_t addresses_per_block() const {
+        return block_size() / sizeof(u32);
+    }
+    size_t group_descriptor_size() const {
+        if (incompatible_features & ext4_feature_incompat_64bit)
+            return group_descriptor_size_num;
+        else
+            return min_group_descriptor_size;
+    }
+    size_t group_descriptors_per_block() const {
+        return block_size() / group_descriptor_size();
+    }
+};
+
+ 
+// The new version of the directory entry.  Since EXT2 structures are
+// stored in intel byte order, and the name_len field could never be
+// bigger than 255 chars, it's safe to reclaim the extra byte for the
+// file_type field.
+
+struct DirEntry {
+    u32 inode;    // Inode number
+    u16 rec_len;  // Directory entry length
+    u8  name_len; // Name length
+    u8  file_type;
+    char name[max_name_len]; // Filename
+};
+
+// Ext2 directory file types.  Only the low 3 bits are used.  The
+// other bits are reserved for now.
+enum class FileType : u8 {
+    Unknown,   // Unknown file type
+    RegFile,   // Regular old file
+    Directory, // Directory
+    CharDev,   // Character device
+    BlockDev,  // Block device
+    FIFO,      // FIFO
+    Socket,    // Socket
+    SymLink,   // Symbolic link
+    Max        // The maximum file type plus one.
+};
+
+// EXT2_DIR_PAD defines the directory entries boundaries
+// Note that it must be a multiple of 4
+constexpr size_t dir_round = (4 - 1);
+
+// Note: T must be a number type
```

Comment:
We can solve this programatically with `Integral`

Reply:
Do you mind explaining a bit more?